pkg/gardenlet/controller/vpaevictionrequirements/reconciler.go

```diff
@@ -0,0 +1,175 @@
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	"github.com/gardener/gardener/pkg/controllerutils"
+	"github.com/gardener/gardener/pkg/gardenlet/apis/config"
+	"github.com/gardener/gardener/pkg/utils/timewindow"
+)
+
+var cpuAndMemory = []corev1.ResourceName{corev1.ResourceMemory, corev1.ResourceCPU}
+var upscaleOnlyRequirement = []*vpaautoscalingv1.EvictionRequirement{
+	{
+		Resources:         cpuAndMemory,
+		ChangeRequirement: vpaautoscalingv1.TargetHigherThanRequests,
+	},
+}
+
+func removeAllEvictionRequirements(vpa *vpaautoscalingv1.VerticalPodAutoscaler) {
+	vpa.Spec.UpdatePolicy.EvictionRequirements = []*vpaautoscalingv1.EvictionRequirement{}
+}
+
+func addDenyDownscalingEvictionRequirement(vpa *vpaautoscalingv1.VerticalPodAutoscaler) {
+	vpa.Spec.UpdatePolicy.EvictionRequirements = upscaleOnlyRequirement
+}
+
+// Reconciler implements the reconciliation logic for adding/removing EvictionRequirements to VPA objects.
+type Reconciler struct {
+	Config     config.VPAEvictionRequirementsControllerConfiguration
+	SeedClient client.Client
+	Clock      clock.Clock
+}
+
+// Reconcile implements the reconciliation logic for adding/removing EvictionRequirements to VPA objects.
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := logf.FromContext(ctx)
+	seedCtx, cancel := controllerutils.GetMainReconciliationContext(ctx, controllerutils.DefaultReconciliationTimeout)
+	defer cancel()
+
+	vpa := &vpaautoscalingv1.VerticalPodAutoscaler{}
+	err := r.SeedClient.Get(seedCtx, request.NamespacedName, vpa)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("error retrieving object from store: %w", err)
+	}
+
+	log.Info("Reconciling vpa")
+
+	// double check just for fun: does the vpa have our label?
+	if !metav1.HasLabel(vpa.ObjectMeta, constants.LabelVPAEvictionRequirementDownscaleRestriction) {
+		err := fmt.Errorf("label %q not found, although marker label %q is present", constants.LabelVPAEvictionRequirementDownscaleRestriction, constants.LabelVPAEvictionRequirementsController)
+		log.Error(err, "Error while parsing the label value:")
+		return reconcile.Result{}, err
+	}
+	value := vpa.GetLabels()[constants.LabelVPAEvictionRequirementDownscaleRestriction]
+	log.Info("Found the label "+constants.LabelVPAEvictionRequirementDownscaleRestriction, "value", value)
+	switch value {
+	case constants.EvictionRequirementNever:
+		return r.reconcileVPAForDownscaleDisabled(ctx, vpa)
+	case constants.EvictionRequirementInMaintenanceWindowOnly:
+		return r.reconcileVPAForDownscaleInMaintenanceOnly(ctx, vpa)
+	default:
+		err := fmt.Errorf("unsupported label value found: %q, supported are only %q and %q", value, constants.EvictionRequirementNever, constants.EvictionRequirementInMaintenanceWindowOnly)
+		log.Error(err, "Error while parsing the label value:")
+		return reconcile.Result{}, err
```

Comment:
Does it really make sense to return an error here (meaning, that the controller retries again and again)? It will never succeed, so better just log the error but don't return it?

Reply:
done.